File: compiler/bench/queens.c
// expected: 92
#include <stdio.h>
#include <stdlib.h>
int board[8];
long solutions = 0;

int safe(int row, int col) {
    for (int i = 0; i < col; i++) {
        if (board[i] == row) return 0;
        if (abs(board[i] - row) == abs(i - col)) return 0;
    }
    return 1;
}

void queens(int col) {
    if (col == 8) { solutions++; return; }
    for (int row = 0; row < 8; row++) {
        if (safe(row, col)) {
            board[col] = row;
            queens(col + 1);
        }
    }
}

int main() {
    queens(0);
    printf("%ld\n", solutions);
    return 0;
}

File: compiler/bench/queens.fs
\ expected: 92
\ 8-queens count solutions recursively

create board 8 allot
variable solutions

: safe? ( row col -- flag )
  0 begin
    dup 3 pick < while
    dup board + c@
    3 pick = if 2drop drop 0 exit then
    dup board + c@ 3 pick -
    2 pick 3 pick - abs = if 2drop drop 0 exit then
    1+
  repeat 2drop drop 1 ;

: queens ( col -- ) recursive
  dup 8 = if drop 1 solutions +! exit then
  8 0 do
    i over safe? if
      i over board + c!
      dup 1+ recurse
    then
  loop drop ;

: main
  0 solutions !
  0 queens
  solutions @ . cr ;
